ubireader/ubifs/decrypt.py

```diff
@@ -0,0 +1,98 @@
+from ubireader.ubifs.defines import UBIFS_XATTR_NAME_ENCRYPTION_CONTEXT
+from ubireader.debug import error
+from cryptography.hazmat.primitives.ciphers import (
+   Cipher, algorithms, modes
+)
+
+AES_BLOCK_SIZE = algorithms.AES.block_size // 8
+
+def lookup_inode_nonce(inodes: dict, inode: dict) -> bytes:
+    # get the extended attribute 'xent' of the inode
+    if 'xent' not in inode or not inode['xent']:
+        raise ValueError(f"No xent found for inode {inode}")
+
+    for xattr_inode in inode['xent']:
+        if (xattr_inode.name == UBIFS_XATTR_NAME_ENCRYPTION_CONTEXT):
+            nonce_ino = inodes[xattr_inode.inum]['ino']
+            nonce = nonce_ino.data[-16:]
+            if len(nonce) != 16:
+                raise ValueError(f"Invalid nonce length for inode {inode}")
+            return nonce
+    
+
+def derive_key_from_nonce(master_key: bytes, nonce: bytes) -> bytes: 
+    encryptor = Cipher(
+        algorithms.AES(nonce),
+        modes.ECB(),
+    ).encryptor() 
+    derived_key = encryptor.update(master_key) + encryptor.finalize()
+    return derived_key
+
+
+def filename_decrypt(key: bytes, ciphertext: bytes):
+    
+    # using AES CTS-CBC mode not supported by pyca cryptography 
+    if len(ciphertext) > AES_BLOCK_SIZE:
+        # Cipher Text Stealing Step
+        decryptor = Cipher(
+            algorithms.AES(key[:32]),
+            modes.ECB(),
+        ).decryptor() 
+        
+        pad = (AES_BLOCK_SIZE - len(ciphertext) % AES_BLOCK_SIZE) % AES_BLOCK_SIZE
+    
+        if pad > 0:  # Steal ciphertext only if needed (CTS)
+            second_to_last = ciphertext[-2*AES_BLOCK_SIZE+pad:-AES_BLOCK_SIZE+pad]
+            plaintext = decryptor.update(second_to_last) + decryptor.finalize()
+            # Apply padding 
+            ciphertext += bytes(plaintext[-pad:])
+        # Swap the last two blocks
+        ciphertext = ciphertext[:-2*AES_BLOCK_SIZE] + ciphertext[-AES_BLOCK_SIZE:] + ciphertext[-2*AES_BLOCK_SIZE:-AES_BLOCK_SIZE]
+    
+    # AES-CBC step
+    NULL_IV = b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+
+    decryptor = Cipher(
+        algorithms.AES(key[:32]),
+        modes.CBC(NULL_IV),
+    ).decryptor()
+    plaintext = decryptor.update(ciphertext) + decryptor.finalize()
+    return plaintext.rstrip(b'\x00')
```

Comment:
This is the point, where `.decode()` could be called, so as a `str` could be returned.
Are there any potential filenames, that are not utf-8 encoded?